### PR TITLE
Migrate OC_Mail to SwiftMailer

### PR DESCRIPTION
**Pull request by LukasReschke**
This refactors OC_Mail to use SwiftMailer instead of PHPMail, furthermore this fixes the unit tests when running locally due to being able to mock the  mailing class. (#12014)

Requires https://github.com/owncloud/3rdparty/pull/143

@PVince81 @th3fallen @bantu @DeepDiver1975 Please review carefully. - Thanks!
# Todos
- [x] Revert #12489

### CODE REVIEW

**Comment by nickvergessen** [comment]
Another problem with PHPMailer:
https://github.com/owncloud/core/issues/13252

**Comment by LukasReschke** [comment]
Would love to get some momentum on this. I rebased it and added some more unit tests for code that wasn't unit-testable before. We need this as well for one of our blue things ;-)

**Comment by LukasReschke** [comment]
Usage is now like: 

```php
$message = $this->mailer->createMessage();
$message->setTo([$email => $username]);
$message->setSubject($subject);
$message->setHtmlBody($mailContent);
$message->setPlainBody($plainTextMailContent);
$message->setFrom([$this->fromMailAddress => $this->defaults->getName()]);
$this->mailer->send($message);
```

**Comment by LukasReschke** [comment]
This is also crucial to be able to test our code… Currently it's impossible to write unit tests for anything that uses OC_Mail… :(

File: core/lostpassword/controller/lostcontroller.php [hunk 1]
```diff
@@ -201,7 +201,7 @@ protected function sendEmail($user) {
 
 		try {
 			// FIXME: should be added to the container and injected in here
-			\OC_Mail::send(
+			\OCP\Util::sendMail(
```
Comment (th3fallen) [inline core/lostpassword/controller/lostcontroller.php]: Erm... This is depreciated and not using your new class...

Comment (LukasReschke) [inline core/lostpassword/controller/lostcontroller.php]: Fair point. Will adjust as well and add unit tests then as well.

Comment (LukasReschke) [inline core/lostpassword/controller/lostcontroller.php]: Done with https://github.com/owncloud/core/commit/bff946cbdf38feafac30c46543ba828d720e1262

**Comment by MorrisJobke** [comment]
@LukasReschke Have a look at scrutinizer ;)

**Comment by LukasReschke** [comment]
@MorrisJobke Why?

https://scrutinizer-ci.com/g/owncloud/core/inspections/c31dd76c-3f2d-4bc3-9a65-02fe110139a3/issues/?status=new&selectedAuthors%5B0%5D=lukas%40owncloud.com&orderField=path&order=asc

Those "bugs" are caused by old code which was copied 1:1 I don't see any reason to check if a template actually succeeded or not. If it fails it fails and it shouldn't (especially since the unit tests are testing this anyways)

**Comment by LukasReschke** [comment]
And Code Duplication in the tests… Well there is worse ;-)

**Comment by DeepDiver1975** [comment]
I agree to the current interface design - need to find the time to review ....

**Comment by MorrisJobke** [comment]
https://scrutinizer-ci.com/g/owncloud/core/inspections/c31dd76c-3f2d-4bc3-9a65-02fe110139a3/issues/files/core/lostpassword/controller/lostcontroller.php?status=new&selectedAuthors%5B0%5D=lukas%40owncloud.com&orderField=path&order=asc ;)

**Comment by LukasReschke** [comment]
> Those "bugs" are caused by old code which was copied 1:1 I don't see any reason to check if a template actually succeeded or not. If it fails it fails and it shouldn't (especially since the unit tests are testing this anyways)

**Comment by MorrisJobke** [comment]
Okay :P

**Comment by LukasReschke** [comment]
:innocent:

**Comment by LukasReschke** [comment]
@DeepDiver1975 Found anytime to review? – Rebasing does not really make sense before you have found time as I don't want to rebase this for the 40th time ;-)

File: lib/public/mail/util.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * @package OCP\Mail
+ */
+class Util {
+	/**
+	 * Checks if an e-mail address is valid
+	 *
+	 * @param string $email Email address to be validated
+	 * @return bool True if the mail address is valid, false otherwise
+	 */
+	public static function validateMailAddress($email) {
```
Comment (DeepDiver1975) [inline lib/public/mail/util.php]: bääh - static Kung-Fu - can we keep this within the Mailer class - does that make sense?

File: tests/lib/mail.php [hunk 3]
```diff
@@ -1,53 +0,0 @@
...
-		$actual = \OC_Mail::buildAsciiEmail($address);
-		$this->assertEquals($expected, $actual);
-	}
-
-	public function buildAsciiEmailProvider() {
-		return array(
-			array('info@example.com', 'info@example.com'),
-			array('info@xn--cjr6vy5ejyai80u.com', 'info@國際化域名.com'),
-			array('info@xn--mller-kva.de', 'info@müller.de'),
-			array('info@xn--mller-kva.xn--mller-kva.de', 'info@müller.müller.de'),
```
Comment (DeepDiver1975) [inline tests/lib/mail.php]: hey - I want my müller-müller address back ;-)

**Comment by DeepDiver1975** [comment]
@LukasReschke you have my :+1: on this - did ever anybody really send an email with this pr?

:hide:

File: lib/public/mail/util.php (hunk 2 above)
Comment (LukasReschke) [inline lib/public/mail/util.php]: Sure. Will move it there.

**Comment by LukasReschke** [comment]
> did ever anybody really send an email with this pr?

Real men don't test their changes ;-)

Yes – tested it locally – though a while ago and probably 5 rebases ago. Worked firn at this time ;-)

**Comment by DeepDiver1975** [comment]
> Real men don't test their changes ;-)

**Comment by MorrisJobke** [comment]
Ping me once it is rebased and I will review and test with real mails ;)

**Comment by LukasReschke** [comment]
@MorrisJobke Done… I'm pretty sure I fucked up some 3rdparty ref so if you could update it this would be awesome *hides* ;-)

*****PING****** ;-)

**Comment by DeepDiver1975** [comment]
> @MorrisJobke Done… I'm pretty sure I fucked up some 3rdparty ref so if you could update it this would be awesome hides ;-)
> 
> **PING*** ;-)

let me have a closer look ...

**Comment by DeepDiver1975** [comment]
@LukasReschke I took the freedom to rebase the branch

**Comment by DeepDiver1975** [comment]
I just tried to connect with two different smtp servers .... without success

error message somewhere in the depth:
````
Expected response code 250 but got code "", with message ""

**Comment by LukasReschke** [comment]
What did you configure exactly?

**Comment by LukasReschke** [comment]
So, in some cases `SwiftMailer` seems to have a problem connecting to the remote host when the connection hasn't been started manually.

I added this and modified the email test wizard as well to show a more technical error message in case something happens.

**Comment by LukasReschke** [comment]
If you want to try with your GMail account you might need to allow login in https://www.google.com/settings/security/lesssecureapps?pli=1 as well

**Comment by LukasReschke** [comment]
And another round of rebase required… – @DeepDiver1975 Let me know if the current approach works then I'll rebase :see_no_evil:

**Comment by MorrisJobke** [comment]
@LukasReschke @DeepDiver1975 Lets get this in. Feel free to ping me, we need this.

**Comment by DeepDiver1975** [comment]
@MorrisJobke we have to ensure that existing functional mail setting continue to work with swiftmailer - I failed to make this work the other day ......

**Comment by DeepDiver1975** [comment]
Testplan:
- [ ] use google smtp
- [ ] use owncloud company smtp
- [ ] use some self hosted smtp with selfsigned cert

**Comment by MorrisJobke** [comment]
I can test the first two, but I don't have a SMTP host with a self-signed cert.

**Comment by MorrisJobke** [comment]
@LukasReschke Rebase fun?

**Comment by LukasReschke** [comment]
Rebased. Hopefully all went right…

What works for me:

- SMTP with Gmail
- SMTP with oc.com SMTP

Didn't test with this checkout:
- Plain PHP + sendmail

**Comment by MorrisJobke** [comment]
* [ ] HTML/Plain mails are shown wrong. It seems that there happens something that bricks the multipart mail

**Comment by MorrisJobke** [comment]
BTW: this worked before. Also there is now no log entry anymore, that a mail was send (debug level)

**Comment by LukasReschke** [comment]
>  HTML/Plain mails are shown wrong. It seems that there happens something that bricks the multipart mail

Works fine here. So there is nothing left than:



Used code:
```php

	$mailer = \OC::$server->getMailer();
	$message = $mailer->createMessage();
	$message->setTo(['lukas@owncloud.com']);
	$message->setPlainBody('My plain body');
	$message->setHtmlBody('<strong>My HTML Body</strong>');
	$mailer->send($message);
```

It also works fine when only sending an HTML or Plain Body…

**Comment by LukasReschke** [comment]
> BTW: this worked before. Also there is now no log entry anymore, that a mail was send (debug level)

Log message is back.

**Comment by MorrisJobke** [comment]
@LukasReschke Try to send the share link via mail:

Before
bildschirmfoto von 2015-03-16 13-48-33
(the image is broken because it's a localhost URL)

After
bildschirmfoto von 2015-03-16 13-48-10

This is the same with Thunderbird and Google Mail. So I would say an problem at our side rather than the mail client ;)

**Comment by LukasReschke** [comment]
Should be fixed. That was caused by the legacy `OCP\Util::sendMail` which had a really confusing signature and I was not sure what `$altbody` was for. (I assumed it would be the HTML text while it is the HTML one)

**Comment by MorrisJobke** [comment]
Works now :+1:

**Comment by nickvergessen** [comment]
:+1: tested with the activity emails and added some html to them.

**Pull request merged**